### PR TITLE
Introducing Rewrite which implements DataFrame constructor

### PR DESCRIPTION
**Pull request by AlexanderKalistratov**
Introducing **rewrite** which replaces `DataFrame({'A': a, 'B': b})` with `init_dataframe(a, b, index, 'A', 'B')`
This is a temporally solution and the only reason it is existed is lack of support in Numba of literal lists and dicts (and dicts of lists?). And should be removed as soon as such support appears.

All current tests should works in the same way as before (except few tests which unexpectedly get broken due to unused columns elimination stopped to work)

Introduced new test: **test_create_string_index**. This test wouldn't work without **rewrite**

### CODE REVIEW

File: sdc/rewrites/dataframe_constructor.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+
+    def apply(self):
+        init_df_stmt = import_function(init_dataframe, self._block, self._func_ir)
+
+        for stmt, extract, cleanup in self._calls_to_rewrite:
+            args = get_call_parameters(call=stmt.value, arg_names=self._df_arg_list)
+
+            old_data = args['data']
+            old_columns = args['columns'] if 'columns' in args else None
```
Comment (densmirn) [inline sdc/rewrites/dataframe_constructor.py]: Looks like this line is equal to `old_columns = args.get('columns')`.

File: sdc/rewrites/dataframe_constructor.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+        return data_args, columns_args
+
+
+    @staticmethod
+    def _extract_dict_args(args, block, func_ir):
+        dict_items = get_dict_items(args['data'], block, func_ir)
+
+        data_args = [item[1] for item in dict_items]
+        columns_args = [item[0] for item in dict_items]
```
Comment (densmirn) [inline sdc/rewrites/dataframe_constructor.py]: Probably you can use `dict_items.keys()` and `dict_items.values()` here.

File: sdc/rewrites/dataframe_constructor.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+
+    @staticmethod
+    def _replace_call(stmt, new_call, args, block, func_ir):
+        func = stmt.value
+
+        data_args = args['data']
+        columns_args = args['columns']
+
+        index_args = args['index'] if 'index' in args else None
```
Comment (densmirn) [inline sdc/rewrites/dataframe_constructor.py]: `index_args = args.get('index')`

File: sdc/rewrites/dataframe_constructor.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+                if is_list(var, block, func_ir):
+                    call_stmt = make_call(fix_df_array, [var], {}, block, func_ir, var.loc)
+                    insert_before(block, call_stmt, stmt)
+                    new_args.append(call_stmt.target)
+                else:
+                    new_args.append(var)
+
+            return new_args
+
+        raise ValueError()
```
Comment (densmirn) [inline sdc/rewrites/dataframe_constructor.py]: Do not we want to add message for the exception?

File: sdc/rewrites/dataframe_constructor.py (hunk 2 above)
Comment (AlexanderKalistratov) [reply sdc/rewrites/dataframe_constructor.py]: Unfortunately no. dict_items is list of tuples, not dict

File: sdc/rewrites/dataframe_constructor.py (hunk 3 above)
Comment (AlexanderKalistratov) [reply sdc/rewrites/dataframe_constructor.py]: Thanks, fixed

File: sdc/rewrites/dataframe_constructor.py (hunk 4 above)
Comment (AlexanderKalistratov) [reply sdc/rewrites/dataframe_constructor.py]: Completely removed

File: sdc/rewrites/dataframe_constructor.py (hunk 1 above)
Comment (AlexanderKalistratov) [reply sdc/rewrites/dataframe_constructor.py]: Thanks, fixed

**Comment by AlexanderKalistratov** [comment]
@shssf @kozlov-alexey please pay attention. This PR is disabling workaround with lightweight passes and skips a bunch of tests

**Comment by AlexanderKalistratov** [comment]
@shssf are you ok how workaround is disabled? It is just commented. Are you fine with this? Or do you want to have special config parameter?

**Pull request merged**